flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/TemporalTableFunction.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.table.expressions.Expression;
+import org.apache.flink.table.operations.TableOperation;
+import org.apache.flink.types.Row;
+
+import java.sql.Timestamp;
+
+/**
+ * Class representing temporal table function over some history table. A
+ * {@link TemporalTableFunction} is also an instance of {@link TableFunction}.
+ *
+ * <p>Currently {@link TemporalTableFunction}s are only supported in streaming.
+ */
+@PublicEvolving
+public class TemporalTableFunction extends TableFunction<Row> {
+
+	private final transient TableOperation underlyingHistoryTable;
+	private final transient Expression timeAttribute;
+	private final transient Expression primaryKey;
+	private final RowTypeInfo resultType;
+
+	private TemporalTableFunction(
+			TableOperation underlyingHistoryTable,
+			Expression timeAttribute,
+			Expression primaryKey,
+			RowTypeInfo resultType) {
+		this.underlyingHistoryTable = underlyingHistoryTable;
+		this.timeAttribute = timeAttribute;
+		this.primaryKey = primaryKey;
+		this.resultType = resultType;
+	}
+
+	public void eval(Timestamp row) {
+		throw new IllegalStateException("This should never be called");
+	}
+
+	public Expression getTimeAttribute() {
+		return timeAttribute;
+	}
+
+	public Expression getPrimaryKey() {
+		return primaryKey;
+	}
+
+	@Override
+	public RowTypeInfo getResultType() {
+		return resultType;
+	}
+
+	public TableOperation getUnderlyingHistoryTable() {
+		if (underlyingHistoryTable == null) {
+			throw new IllegalStateException("Accessing table field after planing/serialization");
+		}
+		return underlyingHistoryTable;
+	}
+
+	public static TemporalTableFunction create(
```

Comment:
can we make this package private or avoid that users use this method? The least we can do is marking it `@Internal` but this is still not perfect. The `Impl` approach was nicer.

Reply:
We can still go with the `Impl` approach. Will update it.